lib/src/kratos_client.dart

```diff
@@ -1153,69 +1239,107 @@ class KratosClient {
         'password': password,
+        'csrf_token': settingsFlow.csrfToken,
       }),
       headers: _buildHeaders({'X-Session-Token': kratosToken}),
     );
 
-    return switch (settingsFlow.statusCode) {
+    return switch (settingsResponse.statusCode) {
       200 => UpdateSuccess(),
       403 => UpdateRequiresReauthorization(),
-      400 => UpdateFailure(error: _handleChangePasswordError(settingsFlow)),
+      400 => UpdateFailure(error: _handleChangePasswordError(settingsResponse)),
       _ => UpdateFailure(error: null),
     };
   }
 
-  Future<UserProfile> getUserProfile() async {
+  Future<SessionResult> getSession() async {
     final kratosToken = await _credentialsStorage.read();
+    if (!kIsWeb && kratosToken == null) {
+      return const SessionErrorResult();
+    }
+
+    try {
+      final whoamiResponse = await _client.get(
+        _buildUri(path: 'sessions/whoami'),
+        headers: _buildHeaders({'X-Session-Token': kratosToken}),
+      );
+
+      if (whoamiResponse.statusCode == 200) {
+        return SessionSuccessResult(
+          Session.fromJson(
+            json.decode(whoamiResponse.body) as Map<String, dynamic>,
+          ),
+        );
+      }
+
+      return const SessionErrorResult();
+    } catch (e, st) {
+      _logger.warning('Error getting session', e, st);
+      return const SessionErrorResult();
+    }
+  }
 
-    if (kratosToken == null) {
+  Future<UserProfile> getUserProfile() async {
+    final sessionResult = await getSession();
+    if (sessionResult is! SessionSuccessResult) {
       return ErrorGettingUserProfile();
     }
 
-    final whoamiResponse = await _client.get(
-      _buildUri(path: 'sessions/whoami'),
-      headers: _buildHeaders({'X-Session-Token': kratosToken}),
+    final session = sessionResult.session;
+
+    final userId = session.identity.id;
+    final traits = session.identity.traits;
+    final profileTraits = traits.entries
+        .map(
+          (e) => ProfileTrait(
+            traitName: e.key,
+            value: e.value,
+          ),
+        )
+        .toList();
+    return UserProfileData(
+      traits: profileTraits,
+      userId: userId,
     );
+  }
 
-    if (whoamiResponse.statusCode == 200) {
-      try {
-        final session = Session.fromJson(
-          json.decode(whoamiResponse.body) as Map<String, dynamic>,
-        );
-        final userId = session.identity.id;
-        final traits = session.identity.traits;
-        final profileTraits = traits.entries
-            .map(
-              (e) => ProfileTrait(
-                traitName: e.key,
-                value: e.value,
-              ),
-            )
-            .toList();
-        return UserProfileData(
-          traits: profileTraits,
-          userId: userId,
-        );
-      } catch (e, st) {
-        _logger.warning('Error getting user profile', e, st);
-      }
+  Future<SessionValidityResult> isSessionValid() async {
+    if (!kIsWeb) {
```

Comment:
Why can't we have the same code for both?

Reply:
On native we have direct access to a token in credential storage and on web session is store in http-only cookie that is not accessible on a client. So on native we could save some time and one request for every validity check. But theoretically session could be invalidated by server, so we could probably unify this logic to ensure proper validity check

Reply:
Unified